File: Bank.cpp
#include "AccountBuilder.cpp"
#include "AbstractFactory.cpp"
#include "DebitBillFactory.cpp"
#include "CreditBillFactory.cpp"
#include "DepositBillFactory.cpp"
#include "Account.cpp"
#include <fstream>
#include <sstream>
#include <iostream>
#include <vector>
#include <string>

const std::string bd_bills = "bd_bills.csv";
const std::string bd_accounts = "bd_accounts.csv";
const std::string bd_history = "bd_history.csv";

// функции для записи в файл

void Bank::AddBillInFile(Bill& bill, int account_id) {
  std::ofstream file;
  file.open(bd_bills, std::ofstream::app);
  file << account_id << ',' << bill.unique_id << ',' << -1 << ',' << bill.money << "\n";
  file.close();
}

void AddAccountInFile(Account& account) {
  std::ofstream file;
  file.open(bd_accounts, std::ofstream::app);
  file << std::to_string(account.account_id) << ',' << account.login << ',' << account.password_hash << ',' 
       << account.name << ',' << account.surname << ',' << account.address << ',' << std::to_string(account.passport_num) << "\n";
  file.close();
}


// функции для получения содержимого файла

std::vector<std::vector<std::string>> ReadBD(std::ifstream& file, int n) {
  std::vector<std::vector<std::string>> vec;
  std::string line;
  while (std::getline(file, line)) {
    std::istringstream iss(line);
    std::string str;
    std::vector<std::string> person_data;
    for (int i = 0; i < n; i++) {
      std::getline(iss, str, ',');
      person_data.push_back(str);
    }
    vec.push_back(person_data);
  }
  return vec;
}

std::vector<std::vector<std::string>> ReadFileBDBills() {
  std::ifstream file(bd_bills);
  return ReadBD(file, 4);
}

std::vector<std::vector<std::string>> ReadFileBDAccounts() {  // исходный вектор содержит следующий вид: айди аккаунта, логин, пароль, имя, фамилия, адресс, номер паспорта
  std::ifstream file(bd_accounts);
  return ReadBD(file, 7);
}

std::vector<std::vector<std::string>> ReadFileBDHistory() {
  std::ifstream file(bd_history);
  return ReadBD(file, 7);
}

// функция для создания аккаунта

void Bank::CreateAccount(std::string login, std::string password, std::string name, std::string surname, std::string address, int passport_num) {
  AccountBuilder account_builder;
  std::hash<std::string> passw_hash; 
  Account account = account_builder.GetAccount(*(this));
  account.account_id = id_for_accounts++;
  account.login = login;
  account.password_hash = std::to_string(passw_hash(password));
  account.name = name;
  account.surname = surname;
  account.address = address;
  account.passport_num = passport_num;
  AddAccountInFile(account);
}



// функция подтверждения входа на сайт

bool Bank::LoginTo(std::string login, std::string password) {
  auto BD_accounts = ReadFileBDAccounts();
  std::hash<std::string> hash;
  for (int i = 0; i < BD_accounts.size(); ++i) {
    if (BD_accounts[i][1] == login) {
      return std::to_string(hash(password)) == BD_accounts[i][2];
    }
  }
  return false;
}



//функция создания счёта

DepositBillFactory depositfactory;
DebitBillFactory debitfactory;
CreditBillFactory creditfactory;

bool Bank::CreateBill(BillType type_of_bill, int account_id) {
  if (type_of_bill == credit) {
    CreditBill newcreditbill;
    newcreditbill = creditfactory.CreateCredit();
    AddBillInFile(newcreditbill, account_id);
  } else if (type_of_bill == debit) {
    DebitBill newdebitbill;
    newdebitbill = debitfactory.CreateDebit();
    AddBillInFile(newdebitbill, account_id);
  } else if (type_of_bill == deposit) {
    DepositBill newdepositbill;
    newdepositbill = depositfactory.CreateDeposit();
    AddBillInFile(newdepositbill, account_id);
  } else {
    std::cout << "CreateBill error" << std::endl;
    return false;
  }
  return true;
}



// функция проверки списка счетов

void Bank::CheckBills(int account_id) {
  std::vector<std::vector<std::string> > vec;
  std::ifstream file(bd_bills);
  std::string line;
  while (std::getline(file, line)) {
    std::istringstream iss(line);
    std::string str;
    std::vector<std::string> undervec;
    for (int i = 0; i < 4; i++) {
      std::getline(iss, str, ',');
      undervec.push_back(str);
    }
    vec.push_back(undervec);
  }
  std::cout << "bill_id, type, amount" << std::endl;
  for (int i = 0; i < vec.size(); ++i) {
    if (vec[i][0] == std::to_string(account_id)) {
      std::cout << vec[i][1] << ' ' << vec[i][2] << ' ' << vec[i][3] + "$" << std::endl; 
    }
  }
}

// функция для добавления операции в историю

void addOperationToHistory(std::string account_id, std::string oper_type, std::string bill_from, std::string bill_to, std::string amount_of_money) {
  std::ofstream file;
  file.open(bd_history, std::ofstream::app);
  file << account_id << "," << oper_type << ',' << bill_from << ',' << bill_to << ',' 
       << amount_of_money << "\n";
  file.close();
}

// функция для проверки истории

void Bank::CheckHistory(int account_id) {
  auto vec = ReadFileBDHistory();
  std::cout << "История операций\n";
  for (int i = 0; i < vec.size(); ++i) {
    if (vec[i][0] == std::to_string(account_id)) {
      std::cout << vec[i][1] << ' ' << vec[i][2] << ' ' << vec[i][3] << ' ' << vec[i][4] << '\n'; 
    }
  }
  std::cout << "Конец\n";
}

// функции для взаимодействия со счетами

void Bank::AddMoney(int account_id, int bill_to, int amount_of_money) {
  auto vec = ReadFileBDBills();
  for (int i = 0; i < vec.size(); ++i) {
    if (vec[i][1] == std::to_string(bill_to)) {
      vec[i][3] = std::to_string(std::stoi(vec[i][3]) + amount_of_money);
    }
  }
  std::ofstream file;
  file.open(bd_bills);
  for (int i = 0; i < vec.size(); ++i) {
    file << vec[i][0] << ',' << vec[i][1] << ',' << vec[i][2] << ',' << vec[i][3] << "\n";
  }
  addOperationToHistory(std::to_string(account_id), "Add money ", "", "to " + std::to_string(bill_to), "+" + std::to_string(amount_of_money));
  file.close();
}

bool Bank::WithdrawMoney(int account_id, int bill_from, int amount_of_money) {
  auto vec = ReadFileBDBills();
  for (int i = 0; i < vec.size(); ++i) {
    if (vec[i][1] == std::to_string(bill_from)) {
      if (std::stoi(vec[i][3]) - amount_of_money < 0) {
        return false;
      }
      vec[i][3] = std::to_string(std::stoi(vec[i][3]) - amount_of_money);
    }
  }
  std::ofstream file;
  file.open(bd_bills);
  for (int i = 0; i < vec.size(); ++i) {
    file << vec[i][0] << ',' << vec[i][1] << ',' << vec[i][2] << ',' << vec[i][3] << "\n";
  }
  file.close();
  addOperationToHistory(std::to_string(account_id), "Withdraw money ", "from " + std::to_string(bill_from), "", "-" + std::to_string(amount_of_money));
  return true;
}

bool Bank::Transaction(int account_id_from, int account_id_to, int bill_from, int bill_to, int amount_of_money) {
  auto vec = ReadFileBDBills();
  for (int i = 0; i < vec.size(); ++i) {
    if (vec[i][1] == std::to_string(bill_from)) {
      if (std::stoi(vec[i][3]) - amount_of_money < 0) {
        return false;
      }
      vec[i][3] = std::to_string(std::stoi(vec[i][3]) - amount_of_money);
    }
  }
  for (int i = 0; i < vec.size(); ++i) {
    if (vec[i][1] == std::to_string(bill_to)) {
      vec[i][3] = std::to_string(std::stoi(vec[i][3]) + amount_of_money);
    }
  }
  std::ofstream file;
  file.open(bd_bills);
  for (int i = 0; i < vec.size(); ++i) {
    file << vec[i][0] << ',' << vec[i][1] << ',' << vec[i][2] << ',' << vec[i][3] << "\n";
  }
  file.close();
  addOperationToHistory(std::to_string(account_id_from), "Transaction ", "from " + std::to_string(bill_from), "to " + std::to_string(bill_to), "-" + std::to_string(amount_of_money));
  addOperationToHistory(std::to_string(account_id_to), "Transaction ", "from " + std::to_string(bill_from), "to " + std::to_string(bill_to), "+" + std::to_string(amount_of_money));
  return true;
}

/*
// тестирование
int main() {
  Bank bank;
  bank.CreateAccount("Paxan", "22848", "Pavel", "Dranov", "Pervomayskaya30k7", 224423); // account_id = 0
  bank.CreateAccount("DimaZ", "43348", "Dima", "Matveev", "Pervomayskaya30k7", 224424); // account_id = 1
  bank.CreateBill(deposit, 0);  // bill id = 200758
  bank.CreateBill(credit, 1);  // bill id = 200759
  bank.CreateBill(debit, 0);  // bill id = 200760
  bank.AddMoney(0, 200758, 1000);
  bank.Transaction(0, 1, 200758, 200759, 300);
  bank.CheckBills(0);
  bank.CheckHistory(0);
  std::cout << bank.LoginTo("Paxan", "22849") << ' ' << bank.LoginTo("Paxan", "22848") << '\n';
}*/

File: bank.py
import csv
import hashlib
from bill import Bill
import account

# Константы для путей к файлам
BD_BILLS = "bd_bills.csv"
BD_ACCOUNTS = "bd_accounts.csv"
BD_HISTORY = "bd_history.csv"
# Функции для работы с файлами
def clear_bd():
    open('bd_bills.csv', 'w').close()
    open('bd_history.csv', 'w').close()
    open('bd_accounts.csv', 'w').close()
def add_account_in_file(account):
    with open('bd_accounts.csv', mode='a', newline='') as file:
        writer = csv.writer(file)
        writer.writerow([account.account_id, account.login, account.password_hash, account.name, account.surname, account.address, account.passport_num])

def add_bill_in_file(bill, account_id):
    with open('bd_bills.csv', mode='a', newline='') as file:
        writer = csv.writer(file)
        writer.writerow([account_id, bill.unique_id, bill.bill_type, bill.money])

def read_bd(file_path, n):
    with open(file_path, mode='r') as file:
        reader = csv.reader(file)
        return [row for row in reader]

def read_file_bd_bills():
    ans = read_bd('bd_bills.csv', 4)
    open('bd_bills.csv', 'w').close()
    return ans

def read_file_bd_accounts():
    return read_bd('bd_accounts.csv', 7)

def read_file_bd_history():
    return read_bd('bd_history.csv', 5)

# Добавление операции в историю
def add_operation_to_history(account_id, oper_type, bill_from, bill_to, amount_of_money):
    with open('bd_history.csv', mode='a', newline='') as file:
        writer = csv.writer(file)
        writer.writerow([account_id, oper_type, bill_from, bill_to, amount_of_money])

# Структура банка
class Bank:
    def __init__(self):
        with open('id_for_accounts.csv', mode='r') as file_acc:
            reader = csv.reader(file_acc)
            self.id_for_accounts = int(next(reader)[0])
        with open('id_for_bills.csv', mode='r') as file_bills:
            reader = csv.reader(file_bills)
            self.id_for_bills = int(next(reader)[0])  # Использование next для получения первой строки и преобразование в int

        self.name = "TegridyBank"
        self.active_account_id = -1
    
    def update_id_for_accounts(self):
        with open('id_for_accounts.csv', mode='w') as file_acc:
            writer = csv.writer(file_acc)
            self.id_for_accounts = str(int(self.id_for_accounts) + 1)
            writer.writerow([self.id_for_accounts])

    def update_id_for_bills(self):
        with open('id_for_bills.csv', mode='w') as file_bills:
            writer = csv.writer(file_bills)
            self.id_for_bills = str(int(self.id_for_bills) + 1)
            writer.writerow([self.id_for_bills])        

    def create_account(self, login, password, name, surname, address, passport_num):
        newaccount = account.Account(self.id_for_accounts, login, hashlib.sha256(password.encode()).hexdigest(), name, surname, address, passport_num)
        accs = read_file_bd_accounts()
        for one_acc in accs:
            if login == str(one_acc[1]):
                return False
        add_account_in_file(newaccount)
        self.update_id_for_accounts()
        return True

    def login_to(self, login, password):
        bd_accounts = read_file_bd_accounts()
        for account in bd_accounts:
            if account[1] == login:
                self.active_account_id = account[0]
                return hashlib.sha256(password.encode()).hexdigest() == account[2]
        return False

    def create_bill(self, bill_type, account_id):
        if bill_type == 'credit' or bill_type == 'debit' or bill_type == 'deposit':
            newbill = Bill(self.id_for_bills, bill_type, 0)
            add_bill_in_file(newbill, account_id)
            self.update_id_for_bills()
        else:
            print("CreateBill error")
            return False
        return True

    def check_bills(self, account_id):
        bills = read_file_bd_bills()
        print("bill_id, type, amount")
        for bill in bills:
            if bill[0] == str(account_id):
                print(f"{bill[1]} {bill[2]} ${bill[3]}") # нужно изменить чтоб выводил пользователю
            newbill = Bill(bill[1], bill[2], bill[3])
            add_bill_in_file(newbill, bill[0])
        pass

    def check_history(self, account_id):
        transactions = read_file_bd_history()
        print("История операций")
        for transaction in transactions:
            if transaction[0] == str(account_id):
                print(f"{transaction[1]} {transaction[2]} {transaction[3]} {transaction[4]}") # нужно изменить чтоб выводил пользователю
        print("Конец")
        pass

    def add_money(self, account_id, bill_to, amount_of_money):
        if amount_of_money < 0:
            return "Некорректная сумма"
        found = False
        bills = read_file_bd_bills()
        for bill in bills:
            if bill[1] == str(bill_to):
                found = True
                bill[3] = str(int(float(bill[3])) + amount_of_money)
            newbill = Bill(bill[1], bill[2], bill[3])
            add_bill_in_file(newbill, bill[0])
        add_operation_to_history(account_id, "Add money", "", bill_to, "+" + str(amount_of_money))
        return found


    def withdraw_money(self, account_id, bill_from, amount_of_money):
        if amount_of_money < 0:
            return False
        bills = read_file_bd_bills()
        isfound = 0
        for bill in bills:
            if bill[1] == str(bill_from):
                isfound = 1
                if (int(float(bill[3])) < amount_of_money):
                    isfound = -1
                else:
                    bill[3] = str(int(float(bill[3])) - amount_of_money)
            newbill = Bill(bill[1], bill[2], bill[3])
            add_bill_in_file(newbill, bill[0])
        add_operation_to_history(str(account_id), "Withdraw money", "", "from " + str(bill_from), "-" + str(amount_of_money))
        return isfound

    def transaction(self, account_id_from, account_id_to, bill_from, bill_to, amount_of_money):
        if amount_of_money < 0:
            return "Некорректная сумма"
        bills = read_file_bd_bills()
        isfound = 404
        if bills:
            for bill in bills:
                if len(bill) >= 4:  # Check if bill has enough elements
                    if bill[1] == str(bill_from) and bill[0] == str(account_id_from):
                        isfound = 1
                        if (int(float(bill[3])) < amount_of_money):
                            isfound = -1
                        else:
                            bill[3] = str(int(float(bill[3])) - amount_of_money)
                    newbill = Bill(bill[1], bill[2], bill[3])
                    add_bill_in_file(newbill, bill[0])
        bills = read_file_bd_bills()
        if isfound == 1:
            isfound = 0
            for bill in bills:
                if bill[1] == str(bill_to):
                    isfound = 1
                    bill[3] = str(int(float(bill[3])) + amount_of_money)
                    add_operation_to_history(str(account_id_to), "transaction", "", "to " + str(bill_to), "+" + str(amount_of_money))
                    add_operation_to_history(str(account_id_from), "transaction", "", "from " + str(bill_from), "+" + str(amount_of_money))
                newbill = Bill(bill[1], bill[2], bill[3])
                add_bill_in_file(newbill, bill[0])
        elif isfound == 404:
            for bill in bills:
                newbill = Bill(bill[1], bill[2], bill[3])
                add_bill_in_file(newbill, bill[0])
        else:
            for bill in bills:
                if bill[1] == str(bill_from):
                        bill[3] = str(int(float(bill[3])) + amount_of_money)
                newbill = Bill(bill[1], bill[2], bill[3])
                add_bill_in_file(newbill, bill[0])
        return isfound

# Тестирование
if __name__ == "__main__":
    '''
    bank = Bank()
    clear_bd()
    bank.create_account("Paxan", "22848", "Pavel", "Dranov", "Pervomayskaya30k7", 224423)
    bank.create_account("DimaZ", "43348", "Dima", "Matveev", "Pervomayskaya30k7", 224424)
    bank.create_bill("deposit", 0)
    bank.create_bill("credit", 228)
    bank.create_bill("debit", 1)
    bank.add_money(1, 0, 1000)
    bank.add_money(1, 0, 1500)
    bank.add_money(0, 0, 1000)
    bank.transaction(0, 1, 0, 1, 300)
    bank.withdraw_money(228, 1, 290)
    bank.check_bills(0)
    bank.check_history(1)
    '''
